core/templates/dev/head/pages/exploration_editor/statistics_tab/LearnerActionRenderService.js

```diff
@@ -0,0 +1,237 @@
+          '. Submitted answer "' + answer + '" in card "' + currentStateName +
+          '".</span>';
+      } else {
+        htmlString =
+          '<span class="oppia-issues-learner-action">' + actionIndex +
+          '. Submitted answer "' + answer + '" and moved to card "' +
+          destStateName + '" after spending ' + timeSpentInStateSecs +
+          ' seconds on card "' + currentStateName + '".</span>';
+      }
+      return htmlString;
+    };
+
+    /**
+     * Renders the correct HTML for the table display for MultipleIncorrect
+     * issue.
+     * @param {list(LearnerAction)} finalBlock.
+     * @returns {string}
+     */
+    var renderLearnerActionsTableForMultipleIncorrectIssue = function(
+        finalBlock) {
+      var index = finalBlock.length - 1;
+      var stateName =
+        finalBlock[index].actionCustomizationArgs.state_name.value;
+
+      var tableHTML =
+        '<table class="oppia-issues-learner-action-table"><tr><th>Answer</th>' +
+        '<th>Feedback</th></tr>';
+      for (var i = 0; i < index; i++) {
+        if (finalBlock[i].actionType !== ACTION_TYPE_ANSWER_SUBMIT) {
+          continue;
+        }
+        var answer =
+          finalBlock[i].actionCustomizationArgs.submitted_answer.value;
+        var feedback = finalBlock[i].actionCustomizationArgs.feedback.value;
+        tableHTML +=
+          '<tr><td>' + answer + '</td><td>' + feedback + '</td></tr>';
+      }
+      tableHTML += '</table>';
+      return tableHTML;
+    };
+
+    /**
+     * Renders the correct HTML for the learner action.
+     * @param {LearnerAction} learnerAction.
+     * @param {int} actionIndex.
+     * @returns {string}
+     */
+    var renderLearnerActionHTML = function(learnerAction, actionIndex) {
+      var actionType = learnerAction.actionType;
+      var custArgs = learnerAction.actionCustomizationArgs;
+      if (actionType === ACTION_TYPE_EXPLORATION_START) {
+        return renderExplorationStartActionHTML(
+          custArgs.state_name.value, actionIndex);
+      } else if (actionType === ACTION_TYPE_EXPLORATION_QUIT) {
+        return renderExplorationQuitActionHTML(
+          custArgs.state_name.value, custArgs.time_spent_state_in_msecs.value,
+          actionIndex);
+      } else if (actionType === ACTION_TYPE_ANSWER_SUBMIT) {
+        interactionId = custArgs.interaction_id.value;
+        if (interactionId === 'Continue') {
+          return renderContinueButtonSubmitActionHTML(
+            custArgs.dest_state_name.value,
+            custArgs.time_spent_state_in_msecs.value, actionIndex);
+        } else {
+          return renderAnswerSubmitActionHTML(
+            custArgs.submitted_answer.value, custArgs.dest_state_name.value,
+            custArgs.time_spent_state_in_msecs.value, custArgs.state_name.value,
+            actionIndex);
+        }
+      }
+    };
+
+    var withinBlockUpperBound = function(blockLength) {
```

Comment:
Add a doc here explaining what it means to be an "upper bound".

Reply:
Done.